File: src/Network/Protocol.hs
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE OverloadedStrings #-}

{-|
Message types for real-time communication in 1vs1 mode:
  * Client → Server: Ready, Attack
  * Server → Client: AttackResult, GameOver, Error
-}

module Network.Protocol
    ( -- * Message types
      ClientMessage(..)
    , ServerMessage(..)
    
    -- * Specific message types
    , ReadyMessage(..)
    , AttackMessage(..)
        , StartMessage(..)
    , AttackResultMessage(..)
    , GameOverMessage(..)
    , PlayerReadyMessage(..)
        , GameStartMessage(..)
    , ErrorMessage(..)
    
    -- * Encoding/Decoding
    , decodeClientMessage
    , encodeServerMessage
    ) where

import Data.Aeson (FromJSON, ToJSON, decode, encode)
import qualified Data.ByteString.Lazy as BL
import Data.Text (Text)
import GHC.Generics (Generic)

import Game.Types (Fleet, Position)

-- ============================================================================
-- Client Messages (Client → Server)
-- ============================================================================

-- | Messages sent from client to server
data ClientMessage
    = ReadyMsg ReadyMessage
    | AttackMsg AttackMessage
    | StartMsg StartMessage
    deriving (Show, Eq, Generic)

instance FromJSON ClientMessage
instance ToJSON ClientMessage

-- | Player ready with fleet placement
data ReadyMessage = ReadyMessage
    { rmPlayerId :: Text
    , rmFleet :: Fleet
    } deriving (Show, Eq, Generic)

instance FromJSON ReadyMessage
instance ToJSON ReadyMessage

-- Example JSON:
-- {
--   "type": "ready",
--   "playerId": "abc123",
--   "fleet": [
--     {"shipType": "Destroyer", "shipPosition": {"posRow": 0, "posCol": 0}, ...},
--     ...
--   ]
-- }

-- | Player attack
data AttackMessage = AttackMessage
    { amPlayerId :: Text
    , amPosition :: Position
    } deriving (Show, Eq, Generic)
-- | Host signals to start (navigate to setup)
data StartMessage = StartMessage
    { smPlayerId :: Text
    } deriving (Show, Eq, Generic)

instance FromJSON StartMessage
instance ToJSON StartMessage


instance FromJSON AttackMessage
instance ToJSON AttackMessage

-- Example JSON:
-- {
--   "type": "attack",
--   "playerId": "abc123",
--   "position": {"posRow": 5, "posCol": 3}
-- }

-- ============================================================================
-- Server Messages (Server → Client)
-- ============================================================================

-- | Messages sent from server to client
data ServerMessage
    = AttackResultMsg AttackResultMessage
    | GameOverMsg GameOverMessage
    | GameStartMsg GameStartMessage
    | PlayerReadyMsg PlayerReadyMessage
    | ErrorMsg ErrorMessage
    deriving (Show, Eq, Generic)
-- | Notify both clients to navigate to setup (pre-game placement)
data GameStartMessage = GameStartMessage
    { gsmRoomId :: Text
    } deriving (Show, Eq, Generic)

instance FromJSON GameStartMessage
instance ToJSON GameStartMessage


instance FromJSON ServerMessage
instance ToJSON ServerMessage

-- | Attack result (broadcast to both players)
data AttackResultMessage = AttackResultMessage
    { armAttacker :: Text        -- Attacker playerId
    , armPosition :: Position    -- Attack position
    , armResult :: Text          -- "miss" | "hit" | "sunk"
    , armShipType :: Maybe Text  -- Ship type if sunk
    , armNextTurn :: Text        -- Next player's turn
    } deriving (Show, Eq, Generic)

instance FromJSON AttackResultMessage
instance ToJSON AttackResultMessage

-- Example JSON:
-- {
--   "type": "attack_result",
--   "attacker": "abc123",
--   "position": {"posRow": 5, "posCol": 3},
--   "result": "hit",
--   "shipType": null,
--   "nextTurn": "abc123"
-- }

-- | Game over (broadcast to both players)
data GameOverMessage = GameOverMessage
    { gomWinner :: Text          -- Winner playerId
    , gomWinnerName :: Text      -- Winner name
    , gomReason :: Text          -- "all_ships_sunk" | "opponent_disconnected"
    } deriving (Show, Eq, Generic)

instance FromJSON GameOverMessage
instance ToJSON GameOverMessage
-- | Player ready status update (broadcast to both players)
data PlayerReadyMessage = PlayerReadyMessage
    { prmPlayerId :: Text
    , prmReady :: Bool
    } deriving (Show, Eq, Generic)

instance FromJSON PlayerReadyMessage
instance ToJSON PlayerReadyMessage


-- Example JSON:
-- {
--   "type": "game_over",
--   "winner": "abc123",
--   "winnerName": "Player1",
--   "reason": "all_ships_sunk"
-- }

-- | Error message
data ErrorMessage = ErrorMessage
    { emError :: Text
    , emDetails :: Maybe Text
    } deriving (Show, Eq, Generic)

instance FromJSON ErrorMessage
instance ToJSON ErrorMessage

-- Example JSON:
-- {
--   "type": "error",
--   "error": "Invalid fleet placement",
--   "details": "Ships overlap"
-- }

-- ============================================================================
-- Encoding/Decoding Helpers
-- ============================================================================

-- | Decode client message from JSON ByteString
decodeClientMessage :: BL.ByteString -> Maybe ClientMessage
decodeClientMessage = decode

-- | Encode server message to JSON ByteString
encodeServerMessage :: ServerMessage -> BL.ByteString
encodeServerMessage = encode

File: client/scripts/websocket/protocol.js
/**
 * WebSocket message protocol constants
 * Must match backend Network/Protocol.hs
 */

const Protocol = {
  // ============================================================================
  // Message Types (Client → Server)
  // ============================================================================

  CLIENT_MSG: {
    READY: "ready", // Player ready with fleet placement
    ATTACK: "attack", // Player attack
    START: "start", // Host signals to start setup
  },

  // ============================================================================
  // Message Types (Server → Client)
  // ============================================================================

  SERVER_MSG: {
    ATTACK_RESULT: "attack_result", // Attack result (broadcast to both)
    GAME_OVER: "game_over", // Game ended
    GAME_START: "game_start", // Navigate both players to setup phase
    PLAYER_READY: "player_ready", // Player ready status update
    ERROR: "error", // Error message
  },

  // ============================================================================
  // Result Types
  // ============================================================================

  RESULT: {
    MISS: "miss",
    HIT: "hit",
    SUNK: "sunk",
  },

  // ============================================================================
  // Message Builders (Client → Server)
  // ============================================================================

  /**
   * Build ready message
   * @param {string} playerId
   * @param {Array} fleet - Array of 5 ships
   * @returns {string} JSON string
   */
  buildReadyMessage(playerId, fleet) {
    // Match Haskell Aeson default encoding for ClientMessage
    return JSON.stringify({
      tag: "ReadyMsg",
      contents: {
        rmPlayerId: playerId,
        rmFleet: fleet,
      },
    });
  },

  /**
   * Build attack message
   * @param {string} playerId
   * @param {{posRow: number, posCol: number}} position
   * @returns {string} JSON string
   */
  buildAttackMessage(playerId, position) {
    // Match Haskell Aeson default encoding for ClientMessage
    return JSON.stringify({
      tag: "AttackMsg",
      contents: {
        amPlayerId: playerId,
        amPosition: position,
      },
    });
  },

  /**
   * Build start message (host only)
   * @param {string} playerId
   * @returns {string} JSON string
   */
  buildStartMessage(playerId) {
    // Match Haskell Aeson default encoding for ClientMessage
    return JSON.stringify({
      tag: "StartMsg",
      contents: {
        smPlayerId: playerId,
      },
    });
  },

  // ============================================================================
  // Message Parsers (Server → Client)
  // ============================================================================

  /**
   * Parse incoming WebSocket message
   * @param {string} data - JSON string from server
   * @returns {object|null} Parsed message or null on error
   */
  parseMessage(data) {
    try {
      const raw = JSON.parse(data);
      if (!raw || typeof raw !== "object") return null;

      // If server already sends { type: '...' } shape, keep as-is
      if (Object.prototype.hasOwnProperty.call(raw, "type")) {
        return raw;
      }

      // Normalize Haskell Aeson default sum encoding: { tag: 'Ctor', contents: { ... } }
      const tag = raw.tag || raw.Tag || raw.constructor;
      const contents = raw.contents || raw.Contents || raw;
      switch (tag) {
        case "GameStartMsg":
          return {
            type: this.SERVER_MSG.GAME_START,
            roomId: contents.gsmRoomId,
          };
        case "PlayerReadyMsg":
          return {
            type: this.SERVER_MSG.PLAYER_READY,
            playerId: contents.prmPlayerId,
            ready: contents.prmReady,
          };
        case "AttackResultMsg":
          return {
            type: this.SERVER_MSG.ATTACK_RESULT,
            attacker: contents.armAttacker,
            position: contents.armPosition,
            result: contents.armResult,
            shipType: contents.armShipType,
            nextTurn: contents.armNextTurn,
          };
        case "GameOverMsg":
          return {
            type: this.SERVER_MSG.GAME_OVER,
            winner: contents.gomWinner,
            winnerName: contents.gomWinnerName,
            reason: contents.gomReason,
          };
        case "ErrorMsg":
          return {
            type: this.SERVER_MSG.ERROR,
            error: contents.emError,
            details: contents.emDetails,
          };
        default:
          return raw;
      }
    } catch (error) {
      console.error("Failed to parse WebSocket message:", error);
      return null; // Silent ignore (per requirements)
    }
  },

  /**
   * Check message type
   * @param {object} message
   * @param {string} expectedType
   * @returns {boolean}
   */
  isMessageType(message, expectedType) {
    return message && message.type === expectedType;
  },
};

// Export for use in other modules
if (typeof module !== "undefined" && module.exports) {
  module.exports = Protocol;
}
